File: js/src/memory/embeddings.ts
/**
 * ==============================================================================
 * @termux-ai/chain Embeddings Engine: Zero-Dependency Vectorizers (Node.js ESM)
 * ==============================================================================
 * Provides standard BaseEmbeddings, LocalEmbeddings (llama-server /v1/embeddings),
 * and SparseBM25Embeddings (pure math zero-state offline vectorizer).
 * Zero external heavy dependencies - Pure Node.js 18+ standard library.
 */

export interface BaseEmbeddings {
  embedDocuments(texts: string[]): Promise<number[][]>;
  embedQuery(text: string): Promise<number[]>;
}

export class LocalEmbeddings implements BaseEmbeddings {
  public baseUrl: string;
  public endpoint: string;
  public model: string;
  public timeoutMs: number;
  public batchSize: number;
  public apiKey?: string;

  constructor(options?: {
    baseUrl?: string;
    model?: string;
    timeoutMs?: number;
    batchSize?: number;
    apiKey?: string;
  }) {
    this.baseUrl = (options?.baseUrl ?? "http://127.0.0.1:8080/v1").replace(/\/+$/, "");
    this.endpoint = `${this.baseUrl}/embeddings`;
    this.model = options?.model ?? "default";
    this.timeoutMs = options?.timeoutMs ?? 30000;
    this.batchSize = Math.max(1, options?.batchSize ?? 16);
    this.apiKey = options?.apiKey;
  }

  static local(options?: { endpoint?: string; model?: string }): LocalEmbeddings {
    const ep = options?.endpoint ?? "http://127.0.0.1:8080";
    return new LocalEmbeddings({
      baseUrl: `${ep.replace(/\/+$/, "")}/v1`,
      model: options?.model ?? "bge-micro"
    });
  }

  private async postBatch(texts: string[]): Promise<number[][]> {
    const payload = JSON.stringify({
      model: this.model,
      input: texts
    });

    const headers: Record<string, string> = {
      "Content-Type": "application/json"
    };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const resp = await fetch(this.endpoint, {
        method: "POST",
        headers,
        body: payload,
        signal: controller.signal
      });

      if (!resp.ok) {
        const errText = await resp.text();
        throw new Error(`Embedding request failed (HTTP ${resp.status}): ${errText}`);
      }

      const body: any = await resp.json();
      if (!body.data || !Array.isArray(body.data)) {
        throw new Error("Invalid response schema from embeddings endpoint.");
      }

      const sorted = [...body.data].sort((a, b) => (a.index ?? 0) - (b.index ?? 0));
      return sorted.map((item: any) => {
        const emb = item.embedding;
        if (!Array.isArray(emb) || emb.length === 0) {
          throw new Error("Invalid embedding vector returned by server.");
        }
        return emb;
      });
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async embedDocuments(texts: string[]): Promise<number[][]> {
    if (!texts || texts.length === 0) return [];
    const results: number[][] = [];
    for (let i = 0; i < texts.length; i += this.batchSize) {
      const chunk = texts.slice(i, i + this.batchSize);
      const batchEmbs = await this.postBatch(chunk);
      results.push(...batchEmbs);
    }
    return results;
  }

  async embedQuery(text: string): Promise<number[]> {
    const res = await this.embedDocuments([text]);
    if (!res || res.length === 0) {
      throw new Error("Empty embedding returned for query.");
    }
    return res[0];
  }
}

export class SparseBM25Embeddings implements BaseEmbeddings {
  public dimension: number;
  public k1: number;
  public b: number;
  private docCount: number = 0;
  private docFreq: Map<string, number> = new Map();
  private avgDl: number = 1.0;

  constructor(dimension: number = 256, k1: number = 1.5, b: number = 0.75) {
    this.dimension = dimension;
    this.k1 = k1;
    this.b = b;
  }

  private tokenize(text: string): string[] {
    const tokens = text.toLowerCase().match(/\w+/g) || [];
    return tokens.filter(t => t.length > 1);
  }

  private hashToken(token: string): number {
    let h = 0;
    for (let i = 0; i < token.length; i++) {
      h = (h * 31 + token.charCodeAt(i)) >>> 0;
    }
    return h % this.dimension;
  }

  async embedDocuments(texts: string[]): Promise<number[][]> {
    if (!texts || texts.length === 0) return [];

    const docTokensList = texts.map(t => this.tokenize(t));
    this.docCount += texts.length;
    const totalLen = docTokensList.reduce((acc, toks) => acc + toks.length, 0);
    this.avgDl = Math.max(1.0, totalLen / Math.max(1, docTokensList.length));

    for (const toks of docTokensList) {
      const uniqueTerms = new Set(toks);
      for (const term of uniqueTerms) {
        this.docFreq.set(term, (this.docFreq.get(term) ?? 0) + 1);
      }
    }

    const results: number[][] = [];
    for (const toks of docTokensList) {
      const vec = new Array(this.dimension).fill(0.0);
      if (toks.length === 0) {
        results.push(vec);
        continue;
      }

      const tf = new Map<string, number>();
      for (const t of toks) {
        tf.set(t, (tf.get(t) ?? 0) + 1);
      }

      const dl = toks.length;
      for (const [term, count] of tf.entries()) {
        const slot = this.hashToken(term);
        const df = this.docFreq.get(term) ?? 1;
        const idf = Math.log(1.0 + (this.docCount - df + 0.5) / (df + 0.5));
        const bm25Tf = (count * (this.k1 + 1.0)) / (count + this.k1 * (1.0 - this.b + this.b * (dl / this.avgDl)));
        vec[slot] += Math.max(0.0, idf * bm25Tf);
      }

      const norm = Math.sqrt(vec.reduce((sum, v) => sum + v * v, 0));
      results.push(norm > 0 ? vec.map(v => v / norm) : vec);
    }
    return results;
  }

  async embedQuery(text: string): Promise<number[]> {
    const toks = this.tokenize(text);
    const vec = new Array(this.dimension).fill(0.0);
    if (toks.length === 0) return vec;

    const tf = new Map<string, number>();
    for (const t of toks) {
      tf.set(t, (tf.get(t) ?? 0) + 1);
    }

    const dl = toks.length;
    for (const [term, count] of tf.entries()) {
      const slot = this.hashToken(term);
      const df = this.docFreq.get(term) ?? 1;
      const idf = Math.log(1.0 + (this.docCount - df + 0.5) / (df + 0.5));
      const bm25Tf = (count * (this.k1 + 1.0)) / (count + this.k1 * (1.0 - this.b + this.b * (dl / this.avgDl)));
      vec[slot] += Math.max(0.0, idf * bm25Tf);
    }

    const norm = Math.sqrt(vec.reduce((sum, v) => sum + v * v, 0));
    return norm > 0 ? vec.map(v => v / norm) : vec;
  }
}

File: termux_aichain/memory/embeddings.py
"""
==============================================================================
termux-aichain Embeddings Engine: Zero-Dependency On-Device Vectorizers
==============================================================================
Provides standard BaseEmbeddings, LocalEmbeddings (llama-server /v1/embeddings),
and SparseBM25Embeddings (pure math zero-state offline vectorizer).
Zero external heavy dependencies - Pure Python 3.10+ standard library.
"""

from __future__ import annotations
import re
import math
import json
import urllib.request
import urllib.error
from abc import ABC, abstractmethod
from collections import Counter
from typing import Any, Dict, List, Optional, Sequence


class EmbeddingGenerationError(RuntimeError):
    """Raised when an embedding backend fails to vectorize text."""
    pass


class BaseEmbeddings(ABC):
    """Abstract Base Class for text embeddings interfaces."""

    @abstractmethod
    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        """Vectorizes a sequence of documents."""
        pass

    @abstractmethod
    def embed_query(self, text: str) -> List[float]:
        """Vectorizes a single query text string."""
        pass


class LocalEmbeddings(BaseEmbeddings):
    """
    Connects to on-device llama-server or OpenAI-compatible /v1/embeddings endpoint.
    Zero Python dependencies - Uses standard urllib with strict timeouts and batching.
    """

    def __init__(
        self,
        base_url: str = "http://127.0.0.1:8080/v1",
        model: str = "default",
        timeout_seconds: float = 30.0,
        batch_size: int = 16,
        api_key: Optional[str] = None
    ):
        self.base_url = base_url.rstrip("/")
        self.endpoint = f"{self.base_url}/embeddings"
        self.model = model
        self.timeout_seconds = timeout_seconds
        self.batch_size = max(1, batch_size)
        self.api_key = api_key

    @classmethod
    def local(
        cls,
        endpoint: str = "http://127.0.0.1:8080",
        model: str = "bge-micro",
        **kwargs: Any
    ) -> "LocalEmbeddings":
        """Factory method for local edge llama-server embeddings."""
        v1_url = f"{endpoint.rstrip('/')}/v1"
        return cls(base_url=v1_url, model=model, **kwargs)

    def _post_batch(self, texts: List[str]) -> List[List[float]]:
        payload = {
            "model": self.model,
            "input": texts
        }
        data = json.dumps(payload).encode("utf-8")
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        req = urllib.request.Request(self.endpoint, data=data, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req, timeout=self.timeout_seconds) as resp:
                if resp.status != 200:
                    raise EmbeddingGenerationError(f"Embeddings endpoint returned HTTP status {resp.status}")
                raw = resp.read().decode("utf-8")
                body = json.loads(raw)
        except urllib.error.HTTPError as ex:
            err_body = ex.read().decode("utf-8", errors="ignore") if hasattr(ex, "read") else ""
            raise EmbeddingGenerationError(f"Embedding request failed (HTTP {ex.code}): {err_body or str(ex)}") from ex
        except Exception as ex:
            raise EmbeddingGenerationError(f"Embedding endpoint connection error: {str(ex)}") from ex

        if "data" not in body or not isinstance(body["data"], list):
            raise EmbeddingGenerationError(f"Invalid response schema from embeddings endpoint: {body}")

        # Sort embeddings by original response index
        sorted_items = sorted(body["data"], key=lambda x: x.get("index", 0))
        embeddings: List[List[float]] = []
        for item in sorted_items:
            emb = item.get("embedding", [])
            if not emb or any(math.isnan(v) or math.isinf(v) for v in emb):
                raise EmbeddingGenerationError("Model returned invalid or NaN/Inf vector values.")
            embeddings.append(emb)
        return embeddings

    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        if not texts:
            return []
        all_embeddings: List[List[float]] = []
        for i in range(0, len(texts), self.batch_size):
            chunk = texts[i : i + self.batch_size]
            all_embeddings.extend(self._post_batch(chunk))
        return all_embeddings

    def embed_query(self, text: str) -> List[float]:
        res = self.embed_documents([text])
        if not res:
            raise EmbeddingGenerationError("Empty embedding returned for query.")
        return res[0]


class SparseBM25Embeddings(BaseEmbeddings):
    """
    100% Offline, Zero-State Hashing BM25 / TF-IDF Vectorizer.
    Generates deterministic bounded fixed-dimension sparse float vectors without any model weights.
    """

    def __init__(self, dimension: int = 256, k1: float = 1.5, b: float = 0.75):
        self.dimension = dimension
        self.k1 = k1
        self.b = b
        self._doc_count = 0
        self._doc_freq: Counter[str] = Counter()
        self._avg_dl = 1.0

    def _tokenize(self, text: str) -> List[str]:
        tokens = re.findall(r"\w+", text.lower())
        return [t for t in tokens if len(t) > 1]

    def _hash_token(self, token: str) -> int:
        h = 0
        for ch in token:
            h = (h * 31 + ord(ch)) & 0xFFFFFFFF
        return h % self.dimension

    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        if not texts:
            return []

        doc_tokens_list = [self._tokenize(t) for t in texts]
        self._doc_count += len(texts)
        total_len = sum(len(toks) for toks in doc_tokens_list)
        self._avg_dl = max(1.0, total_len / max(1, len(doc_tokens_list)))

        for toks in doc_tokens_list:
            unique_terms = set(toks)
            self._doc_freq.update(unique_terms)

        results: List[List[float]] = []
        for toks in doc_tokens_list:
            vec = [0.0] * self.dimension
            if not toks:
                results.append(vec)
                continue
            tf = Counter(toks)
            dl = len(toks)
            for term, count in tf.items():
                slot = self._hash_token(term)
                df = self._doc_freq.get(term, 1)
                idf = math.log(1.0 + (self._doc_count - df + 0.5) / (df + 0.5))
                bm25_tf = (count * (self.k1 + 1.0)) / (count + self.k1 * (1.0 - self.b + self.b * (dl / self._avg_dl)))
                vec[slot] += max(0.0, idf * bm25_tf)

            # L2 Normalize
            norm = math.sqrt(sum(v * v for v in vec))
            if norm > 0:
                vec = [v / norm for v in vec]
            results.append(vec)
        return results

    def embed_query(self, text: str) -> List[float]:
        toks = self._tokenize(text)
        vec = [0.0] * self.dimension
        if not toks:
            return vec
        tf = Counter(toks)
        dl = len(toks)
        for term, count in tf.items():
            slot = self._hash_token(term)
            df = self._doc_freq.get(term, 1)
            idf = math.log(1.0 + (self._doc_count - df + 0.5) / (df + 0.5))
            bm25_tf = (count * (self.k1 + 1.0)) / (count + self.k1 * (1.0 - self.b + self.b * (dl / self._avg_dl)))
            vec[slot] += max(0.0, idf * bm25_tf)

        norm = math.sqrt(sum(v * v for v in vec))
        if norm > 0:
            vec = [v / norm for v in vec]
        return vec
